File: src/jp/saka1029/land/Land.java
package jp.saka1029.land;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.TreeMap;

public class Land {
    
    class Region {
        
        final String[] names;
        final double[] distances;
        
        Region(Triangle triangle) {
            this.names = Arrays.copyOf(triangle.points, 3);
            this.distances = new double[3];
            for (int i = 0; i < 3; ++i) {
                String left = names[i];
                String right = i < 2 ? names[i + 1] : names[0];
                Double distance = Land.this.distances.get(left + right);
                if (distance == null)
                    throw new IllegalArgumentException(
                        String.format("%sにおける点%sと点%sの距離が定義されていません", triangle, left, right));
                this.distances[i] = distance;
            }
        }
        
        Point third(double xl, Point a, double al, Point b, double bl) {
            double alpha = Math.acos((al * al + xl * xl - bl * bl) / (2 * al * xl));
            Point xx = b.minus(a).multiply(al / xl);
            return a.plus(xx.rotate(alpha));
        }

        Point third(Map<String, Point> points, String name, double xl, Point a, double al, Point b, double bl) {
            Point p = third(xl, a, al, b, bl);
            points.put(name, p);
            return p;
        }

        Point third(Map<String, Point> points) {
            Point p0 = points.get(names[0]);
            Point p1 = points.get(names[1]);
            Point p2 = points.get(names[2]);
            Point x;
            if (p0 != null && p1 != null)
                x = third(points, names[2], distances[0], p0, distances[2], p1, distances[1]);
            else if (p0 != null && p2 != null)
                x = third(points, names[1], distances[2], p2, distances[1], p0, distances[0]);
            else if (p2 != null && p1 != null)
                x = third(points, names[0], distances[1], p1, distances[0], p2, distances[2]);
            else
                x = null;
            return x;
        }
        
        double area() {
            double s = (distances[0] + distances[1] + distances[2]) / 2.0;
            return Math.sqrt(s * (s - distances[0]) * (s - distances[1]) * (s - distances[2]));
        }

        @Override
        public String toString() {
            return String.format("%s--%s->%s--%s->%s--%s->%s",
                names[0], distances[0],
                names[1], distances[1],
                names[2], distances[2],
                names[0]);
        }
    }

    public final String startBase, endBase;
    public final List<Edge> edges;
    public final List<Triangle> triangles;
    public final Map<String, Double> distances;
    public final Map<String, Point> points;
    public final double minX, maxX, minY, maxY;
    public final double width, height;
    public final double area, legalShapedArea;
    
    /**
     * 
     * @param startBase 左下隅に配置する頂点の名前を指定します。
     * @param endBase   右下隅に配置する頂点の名前を指定します。
     * @param edges     辺の両端の頂点の名前と辺の長さを指定します。
     * @param triangles 三角形の頂点の名前を指定します。
     *                  頂点は右回りに指定する必要があります。
     */
    public Land(String startBase, String endBase, List<Edge> edges, List<Triangle> triangles) {
        if (startBase == null)
            throw new IllegalArgumentException("startBase");
        if (endBase == null)
            throw new IllegalArgumentException("endBase");
        if (edges == null || edges.isEmpty())
            throw new IllegalArgumentException("edges");
        if (triangles == null || triangles.isEmpty())
            throw new IllegalArgumentException("triangles");
        Map<String, Double> distances = new HashMap<>();
        for (Edge e : edges) {
            distances.put(e.left + e.right, e.distance);
            distances.put(e.right + e.left, e.distance);
        }
        if (!distances.containsKey(startBase + endBase))
            throw new IllegalArgumentException("%sと%sの距離がedgesに定義されていません");
        this.distances = Collections.unmodifiableMap(distances);
        this.startBase = startBase;
        this.endBase = endBase;
        this.edges = Collections.unmodifiableList(edges);
        this.triangles = Collections.unmodifiableList(triangles);
        List<Region> regions = new ArrayList<>();
        for (Triangle e : triangles)
            regions.add(new Region(e));
        System.out.println(regions);
        double area = 0;
        for (Region r : regions)
            area += r.area();
        this.area = area;
        System.out.println("total area=" + area);
        Map<String, Point> _points = new TreeMap<>();
        this.points = Collections.unmodifiableMap(_points);
        _points.put(startBase, new Point(0, 0));
        _points.put(endBase, new Point(distances.get(startBase + endBase), 0));
        while (true) {
            int size = regions.size();
            for (Iterator<Region> i = regions.iterator(); i.hasNext(); ) {
                Region a = i.next();
                Point p = a.third(_points);
                if (p != null)
                    i.remove();
            }
            if (regions.size() == size)
                break;
        }
        if (!regions.isEmpty())
            throw new RuntimeException("計算できません。計算済の点=" + points + ", 未計算の領域=" + regions);
        double minX = Double.MAX_VALUE;
        double maxX = Double.MIN_VALUE;
        double minY = Double.MAX_VALUE;
        double maxY = Double.MIN_VALUE;
        for (Point e : _points.values()) {
            minX = Math.min(minX, e.x);
            maxX = Math.max(maxX, e.x);
            minY = Math.min(minY, e.y);
            maxY = Math.max(maxY, e.y);
        }
        this.minX = minX;
        this.maxX = maxX;
        this.minY = minY;
        this.maxY = maxY;
        this.width = maxX - minX;
        this.height = maxY - minY;
        this.legalShapedArea = width * height;
    }
    
    public void writeSVG(File file) throws IOException {
        SVG svg = new SVG(minX, maxX, minY, maxY);
        Point topLeft = new Point(minX, minY);
        Point bottomLeft = new Point(minX, maxY);
        Point topRight = new Point(maxX, minY);
        Point bottomRight = new Point(maxX, maxY);
        svg.polygonDash(Arrays.asList(topLeft, topRight, bottomRight, bottomLeft));
        for (Triangle e : triangles) {
            Point[] ps = new Point[] {points.get(e.points[0]), points.get(e.points[1]), points.get(e.points[2])};
            svg.polygon(Arrays.asList(ps));
        }
        // 辺の長さ
        for (Edge e : edges) {
            String n0 = e.left;
            String n1 = e.right;
            Point p0 = points.get(n0);
            Point p1 = points.get(n1);
            double d = distances.get(n0 + n1);
//            svg.text(p0.plus(p1).multiply(0.5), String.format("%.2f", d));
            svg.text(p0, p1, String.format("%.2f", d));
        }
        // 頂点の名前
        for (Entry<String, Point> e : points.entrySet())
            svg.text(e.getValue(), e.getKey()); 
        // 想定整形地
        double x = maxX - minX;
        double y = maxY - minY;
        svg.text(new Point(minX, minY), new Point(maxX, minY), String.format("%.2f", x));
        svg.text(new Point(minX, minY), new Point(minX, maxY), String.format("%.2f", y));
        svg.writeTo(file);
    }
    
}

File: html/land.js
Point = function(x, y) {
    this.x = x;
    this.y = y;
}

Point.prototype = {
    plus : function(p) {
        return new Point(this.x + p.x, this.y + p.y);
    },

    minus : function(p) {
        return new Point(this.x - p.x, this.y - p.y);
    },

    multiply : function(k) {
        return new Point(this.x * k, this.y * k);
    },

    distance : function(p) {
        var d = this.minus(p);
        return Math.sqrt(d.x * d.x + d.y * d.y);
    },

    rotate : function(a) {
        var sin = Math.sin(a);
        var cos = Math.cos(a);
        return new Point(cos * this.x + sin * this.y, -sin * this.x + cos * this.y);
    },

    toString : function() {
        return "(" + this.x + ", " + this.y + ")";
    },
};

Triangle = function(a, b, c) {
    if (!a) throw "頂点の名前1が定義されていません";
    if (!b) throw "頂点の名前2が定義されていません";
    if (!c) throw "頂点の名前3が定義されていません";
    if (a == b) throw "頂点の名前1(" + a + ")と2(" + b + ")が同じです";
    if (b == c) throw "頂点の名前2(" + b + ")と3(" + c + ")が同じです";
    if (c == a) throw "頂点の名前1(" + c + ")と3(" + a + ")が同じです";
    this.points = [a, b, c];
}

Triangle.prototype = {
    toString: function() {
        return "三角形(" + this.points + ")";
    }
};

Edge = function(left, right, distance) {
    if (!left) throw "頂点の名前1が定義されていません";
    if (!right) throw "頂点の名前2が定義されていません";
    if (!distance) throw "辺の長さが定義されていません";
    if (left == right) throw "頂点の名前1(" + left + ")と2(" + right + ")が同じです";
    this.left = left;
    this.right = right;
    this.distance = distance;
}

Edge.prototype = {
    toString : function() {
        return "辺(" + this.left + ", " + this.right + ", " + this.distance + ")";
    }
};

Region = function(triangle, distances) {
    this.names = triangle.points;
    this.distances = [];
    for (var i = 0; i < 3; ++i) {
        var left = this.names[i];
        var right = i < 2 ? this.names[i + 1] : this.names[0];
        this.distances[i] = distances[left + right];
    }
}

Region.prototype = {
    third2: function(xl, a, al, b, bl) {
        var alpha = Math.acos((al * al + xl * xl - bl * bl) / (2 * al * xl));
        var xx = b.minus(a).multiply(al / xl);
        return a.plus(xx.rotate(alpha));
    },

    third1: function(points, name, xl, a, al, b, bl) {
        var p = this.third2(xl, a, al, b, bl);
        points[name] = p;
        return p;
    },

    third: function(points) {
        var p0 = points[this.names[0]];
        var p1 = points[this.names[1]];
        var p2 = points[this.names[2]];
        var x;
        if (p0 && p1)
            x = this.third1(points, this.names[2], this.distances[0], p0, this.distances[2], p1, this.distances[1]);
        else if (p0 && p2)
            x = this.third1(points, this.names[1], this.distances[2], p2, this.distances[1], p0, this.distances[0]);
        else if (p2 && p1)
            x = this.third1(points, this.names[0], this.distances[1], p1, this.distances[0], p2, this.distances[2]);
        else
            x = null;
        return x;
    },
    
    area: function() {
        var s = (this.distances[0] + this.distances[1] + this.distances[2]) / 2.0;
        var t = s * (s - this.distances[0]) * (s - this.distances[1]) * (s - this.distances[2]);
        if (t < 0)
            throw this.toString() + "は三角形ではありません";
        return Math.sqrt(t);
    },

    toString: function() {
        return "領域(" + this.names[0] + "--" + this.distances[0] + "->"
            + this.names[1] + "--" + this.distances[1] + "->"
            + this.names[2] + "--" + this.distances[2] + "->"
            + this.names[0] + ")";
    }
};

Land = function(startBase, endBase, edges, triangles, /* optional(25) */scale, /* optional(2) */ margin) {
    if (!startBase) throw "基準となる点1の名前が定義されていません";
    if (!endBase) throw "基準となる点2の名前が定義されていません";
    if (startBase == endBase) throw "基準となる点1(" + startBase + ")と点2(" + endBase + ")の名前が同じです";
    this.distances = new Object();
    for (var i in edges) {
        var e = edges[i];
        this.distances[e.left + e.right] = e.distance;
        this.distances[e.right + e.left] = e.distance;
    }
    if (!this.distances[startBase + endBase])
        throw startBase + "と" + endBase + "の距離が定義されていません";
    this.startBase = startBase;
    this.endBase = endBase;
    this.edges = edges;
    this.triangles = triangles;
    try {
        this.calculatePoints();
    } catch (e) {
        if (e == "全体を囲むことができません") {
            var temp = this.startBase;
            this.startBase = this.endBase;
            this.endBase = temp;
            this.calculatePoints();
        } else
            throw e;
    }
    this.minX = Number.MAX_VALUE;
    this.maxX = Number.MIN_VALUE;
    this.minY = Number.MAX_VALUE;
    this.maxY = Number.MIN_VALUE;
    for (var k in this.points) {
        e = this.points[k];
        this.minX = Math.min(this.minX, e.x);
        this.maxX = Math.max(this.maxX, e.x);
        this.minY = Math.min(this.minY, e.y);
        this.maxY = Math.max(this.maxY, e.y);
    }
    this.width = this.maxX - this.minX;
    this.height = this.maxY - this.minY;
    this.legalShapedArea = this.width * this.height;
    this.SCALE = scale === undefined ? 25 : scale;
    this.MARGIN = margin === undefined ? 2 :margin;
}

Land.prototype = {
    scale: function(x) {
        return x * this.SCALE;
    },
    
    transform: function(p) {
        return new Point(
            p.x - this.minX + this.MARGIN,
            this.height - p.y - this.minY + this.MARGIN).multiply(this.SCALE);
    },

    pointsString: function() {
        var r = "";
        for (var i = 0; i < arguments.length; ++i)
            r += arguments[i].x + " " + arguments[i].y + " ";
        return r;
    },

    calculatePoints: function() {
        this.regions = [];
        for (var i in this.triangles) {
            var t = this.triangles[i];
            this.regions.push(new Region(t, this.distances));
        }
        this.area = 0;
        for (var i in this.regions) {
            var r = this.regions[i];
            this.area += r.area();
        }
        this.points = new Object();
        this.points[this.startBase] = new Point(0, 0);
        this.points[this.endBase] = new Point(this.distances[this.startBase + this.endBase], 0);
        while (true) {
            var size = this.regions.length;
            for (var i = 0; i < this.regions.length; ) {
                var r = this.regions[i];
                var p = r.third(this.points);
                if (p) {
                    if (p.y < 0)
                        throw "全体を囲むことができません";
                    this.regions.splice(i, 1);
                } else
                    ++i;
            }
            if (this.regions.length == size)
                break;
        }
        if (this.regions.length != 0)
            throw "計算できません。計算済の点=" + this.points + ", 未計算の領域=" + this.regions;
    },
    
    svg: function() {
        var svg = new SVG(this.scale(this.width + this.MARGIN * 2), this.scale(this.height + this.MARGIN * 2));
        var p = this.transform(new Point(this.minX, this.maxY));
        // 想定整形地
        svg.draw("rect", {
            x: p.x,
            y: p.y,
            width: this.scale(this.width),
            height: this.scale(this.height),
            fill: "none",
            stroke: "black",
            "stroke-width": 2,
            "stroke-dasharray": "20 2",
        });
        // 三角形
        for (var i in this.triangles) {
            var t = this.triangles[i];
            svg.draw("polygon", {
                points: this.pointsString(
                    this.transform(this.points[t.points[0]]), 
                    this.transform(this.points[t.points[1]]), 
                    this.transform(this.points[t.points[2]])),
                fill: "#d0f0d0",
                stroke: "black",
                "stroke-width": 2,
            });
        }
        // 辺の長さ
        for (var i in this.edges) {
            var e = this.edges[i];
            var n0 = e.left;
            var n1 = e.right;
            var p0 = this.points[n0];
            var p1 = this.points[n1];
            var d = this.distances[n0 + n1];
            var center = this.transform(p0.plus(p1).multiply(0.5));
            var diff = p1.minus(p0);
            var alpha = diff.x == 0.0 ? -90 : -Math.atan(diff.y / diff.x) * 180 / Math.PI;
            svg.draw("text", {
                x: center.x,
                y: center.y,
                transform: "rotate(" + alpha + " " + center.x + " " + center.y + ")",
                "font-size": 20,
//              "font-weight": "bold",
                "text-anchor": "middle",
                "dominant-baseline": "text-after-edge"
            }, d.toFixed(2));
        }
        // 頂点の名前
        for (var key in this.points) {
            var p = this.transform(this.points[key]);
            svg.draw("text", {
                x: p.x,
                y: p.y,
                "font-size": 40,
                fill: "red",
                "font-weight": "bold",
                "text-anchor": "middle",
                "dominant-baseline": "middle"
            }, key);
        }
        // 想定整形地 幅
        var p = this.transform(new Point(this.minX, this.minY).plus(new Point(this.maxX, this.minY)).multiply(0.5));
        svg.draw("text", {
            x: p.x,
            y: p.y + 20,
            "font-size": 20,
            "font-weight": "bold",
            "text-anchor": "middle",
            "dominant-baseline": "text-before-edge"
        }, this.width.toFixed(2));
        // 想定整形地 高さ
        var q = this.transform(new Point(this.minX, this.minY).plus(new Point(this.minX, this.maxY)).multiply(0.5));
        svg.draw("text", {
            x: q.x - 20,
            y: q.y,
            "font-size": 20,
            "font-weight": "bold",
            "text-anchor": "middle",
            "dominant-baseline": "text-after-edge",
            transform: "rotate(-90 " + (q.x - 20) + " " + q.y + ")",
        }, this.height.toFixed(2));
        return svg.element;
    }
};
